### ISSUE TITLE
Add API to surface tool to integrate bones into a mesh programmatically

### ISSUE THREAD

**Issue by nvanfleet** [godotengine/godot-proposals#6494]
### Describe the project you are working on

I am working on a 3D project which ultimately requires CylinderMesh instances of varying lengths which are procedurally morphed in appearance which will also need to be twisted in various ways (also procedurally). So in short tentacles of varying size and length in varying positions. 

### Describe the problem or limitation you are having in your project

I would like to be able to create a CylinderMesh, mutate its vertexes (which I have mostly working using SurfaceTool / ArrayMesh) and also bend that cylinder in accordance to some different scenarios. I do not see anyway to programmatically do this. I know the easiest way to do it is to create a skeletalized mesh in Blender and then import it, but the number of bones / length of the cylinder will be extremely variable, as will the radius etc etc which makes me want to control it programmatically.

I noticed that the SurfaceTool and other mesh tools have some functionality for bones using `ArrayMesh.ARRAY_BONES` and `ArrayMesh.ARRAY_WEIGHTS`.

This hypothetically allows for the programmatic (procedural) generation of a skeleton on a mesh, however it appears to not be possible. 

`ArrayMesh.ARRAY_BONES` allows you to link vertexes to bones by index number.
`ArrayMesh.ARRAY_WEIGHTS` allows you to configure weights to those bones and their effect on the vertex.

Skeleton3D allows for you to add bones by name, make a bone the child of another bone, set a resting position or transform on that bone etc.

However there appears to be no way to connect the bones in the mesh, which have no names, to the named bones in the `Skeleton3D`.

If you load a dae file and create a scene from it you get a skeleton with a set of named and childed bones, and if you inspect the mesh it will also have expected `ArrayMesh.ARRAY_BONES` and `ArrayMesh.ARRAY_WEIGHTS`.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I think there needs to be some new APIs which allow for the linking of the bone indexes which are setup in ArrayMesh.ARRAY_BONES and the `Skeleton3D` APIs which is used to create named bones and manipulate them. It appears as if that is currently only possible to make the connection from the Skeleton3D to the mesh via loaded files from a 3D modeling program.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

With my narrow knowledge on the subject it appears as if there is missing API in the ArrayMesh API to allow for the naming of the bones in question, so they can correspond with the bones which are created within the Skeleton3D. 

something like

```
data[ArrayMesh.ARRAY_BONE_NAMES] = ["bone-1", "bone-2", "bone-3", "bone-4", "bone-5", "bone-6"]
```
^
I'm not sure if this is sensible or if there is much data in the mesh itself or the skeleton but this would just auto stitch the indexes of the bones into the mesh.

I'm not sure if this can also be solved with Skeleton3D just having new API to create a named bone taking in a reference to the index of an already existing bone? 

Something like `add_bone(name: String, existingIndex: int)`

Sorry I don't know exactly where the bone name connects to the bone index that is seen in ArrayMesh.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I am not aware of any workarounds. You can pre-create a mesh with bones, but it makes it difficult to work around the varying size, vertex count as well as a varying length / bone count. 

### Is there a reason why this should be core and not an add-on in the asset library?

I am not sure how that would work.

**Comment by nvanfleet** [comment]
I'm not sure if this is useful but I also have a forum post https://godotforums.org/d/33107-procgen-skeletons

**Comment by clayjohn** [comment]
I'm not sure this is necessary. When you add a bone to a Skeleton3D, the bone count is used as the bone ID https://docs.godotengine.org/en/latest/classes/class_skeleton3d.html#class-skeleton3d-method-add-bone

The bone ID is what is used by the mesh. When you programmatically create your list of bones you can keep track of the ID as you add them. Alternatively, you can query the bone ID using ``find_bone()`` https://docs.godotengine.org/en/latest/classes/class_skeleton3d.html#class-skeleton3d-method-find-bone

I think that giving SurfaceTool access to the bone names is going to be too messy in the end and will lead to a lot of difficult interdependencies. 

I think this area is sorely under documented. It's probably worth having a page in the manual describing the relationship between bones, Skeleton3D, and the ArrayMesh from a programming perspective

**Comment by Zireael07** [comment]
> I think this area is sorely under documented. It's probably worth having a page in the manual describing the relationship between bones, Skeleton3D, and the ArrayMesh from a programming perspective

Definitely

**Comment by nvanfleet** [comment]
It's definitely not very documented. I can't so far get it to work in my demo app, I've tried to just add bones after setting up the surface tool. I found one Godot tutorial (that by the way has a texture of a very detailed vagina) and the tutorial seemed partial and out of date anyway.

https://docs.godotengine.org/en/3.0/tutorials/3d/working_with_3d_skeletons.html

This is the linked project and the specific texture in that tutorial btw 
![](https://github.com/slapin/godot-skel3d/blob/master/demo1/images/young_lightskinned_female_diffuse.png)

**Comment by nvanfleet** [comment]
I can attach a snippet somewhere, but is the idea just

1. Add bones and weights into the Array Mesh
2. Add bones using Skeleton3D, their indexes will correspond to the indexes of the array from step 1
3. Configure the name / position of the bones, from there they will manipulate the meshes?

So far I haven't been able to make this work, as in I've done all that and my mesh is unaffected.

**Comment by fire** [comment]
I wasn't able to check closely, but you don't seem to be creating 4 or 8 bones or weights per vertex.

**Comment by nvanfleet** [comment]
@fire I didn't see any documentation about how to add that, but I assumed that one was required and the other might be optional , or does every vertex HAVE TO BE affected by 4 individual bones? It seemed like it was just up to a maximum of 4 unless you activate the scenario where there is 8 active bones per vertex.

Also as of last night my code seems to be causing some errors and maybe the Godot code or my code has a bug. I get `  servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp:622 - Condition "bs > sbs" is true. Continuing.` and sometimes the cylinder displays funny but only after restarting Godot. Sometimes the cylinder isn't created at all. Recently I got an error that implied it still thought it was my old bone setup (I reduced the bone count and it was erroring saying there were 24 instead of 6)

**Comment by nvanfleet** [comment]
I think there is a bug in terms of what order you update the new mesh in the meshInstance / adding bones to the skeleton. I get different buggy behaviors unless I disconnect the Skeleton from the MeshInstance first. Then re-adding the skeleton after creates a new behavior where my cylinder gets really big, presumably due to bones?

**Comment by fire** [comment]
Yes, every vertex HAVE TO BE affected by 4 or 8 individual bones.

**Comment by nvanfleet** [comment]
@fire I am a bit confused as that pr states "every vert needs to be assigned to at least one bone" not that they have to have all 4. Maybe this does work but I am just not knowledgeable enough to know what is happening since it's not documented

**Comment by fire** [comment]
Let me try again.

1. every vertex needs to be assigned one bone and every vertex has 4 or 8 slots for bones.
2. When you have one and only one bone, you may assign 1 bone and a weight > 0, the rest can be random values but weighted 0.

**Comment by nvanfleet** [comment]
@fire yes I was just assigning the one bone and weighting it 1 (because the number is from 0>1.0). I tried to put -1 (to imply empty) in the other slots but it would assert so it seemed like leaving it basically null was working. Maybe that is not correct though

**Comment by fire** [comment]
Bone id -1 is a sentinel value, but picking bone id 0 works. The requirement for the at least one valid bone is for unweighted bones.

**Comment by nvanfleet** [comment]
> https://docs.godotengine.org/en/stable/classes/class_skeleton3d.html#class-skeleton3d-method-register-skin will register a skin for you if you have no idea. I typically don't either for procedural meshes.

Does just doing this after configuring my skeleton work? `self.meshInstance.skin = self.skeleton.create_skin_from_rest_transforms()`

**Comment by nvanfleet** [comment]
Okay I cleaned up my demo code, I think it makes more sense now. But I still get a vertically offset tube if my rest position is set on the bones of the skeleton. 

![Screenshot 2023-03-14 at 4 19 25 PM](https://user-images.githubusercontent.com/1486399/225163333-c40cb759-e645-4e46-bf69-849c140a6905.png)

```
		self.skeleton.add_bone("%s" % i)
		# Set initial resting position based on the assumed bone location and bone increment.
		var bonePosition = bonesIncrement * i - halfHeight
		var transform = Transform3D(Vector3(1, 0, 0), Vector3(0, 1, 0), Vector3(0, 0, 1), Vector3(0, bonePosition, 0))
		self.skeleton.set_bone_rest(i, transform)
```

I'm not sure what space to use, I know that the vertexes end up with negative Y values because the mesh's space is centered in the middle of the mesh. I've tried not offsetting the resting position or offsetting it in different directions and it moves in sort of random directions. 

If I can figure this out I'd love to contribute some updated docs.

```
print("Starting bone count %s" % self.skeleton.get_bone_count())
	
	# Remove any existing bones if they exist.
	if self.skeleton.get_bone_count() > 0:
		print("Cleared bones")
		self.skeleton.clear_bones()
		
	var meshHeight = 6
	var meshRadius = 0.5
	var halfHeight = meshHeight / 2
	var bonesIncrement = 1
	
	# Create mesh
	var newMesh = CylinderMesh.new()
	newMesh.height = meshHeight
	newMesh.bottom_radius = meshRadius
	newMesh.top_radius = meshRadius
	
	var data = newMesh.get_mesh_arrays()
	var vertices = data[ArrayMesh.ARRAY_VERTEX]
	var newBones = PackedInt32Array()
	newBones.resize(vertices.size() * 4)
	var newWeights = PackedFloat32Array()
	newWeights.resize(vertices.size() * 4)
	
	var addedBoneMaxIndex = 0
	for i in vertices.size():
		# Find the nearest bone index based on the bone increment value.
		var vertex = vertices[i]
		var boneAddress = int((floor(vertex.y + halfHeight)/ bonesIncrement))
		
		# Only set one vertex bone to the nearest bone to the vertex. 
		# The rest are already zero'd so they point to bone 0 with a 0 weight.
		var boneLocation = i * 4
		newBones[boneLocation] = boneAddress
		newWeights[boneLocation] = 1
		addedBoneMaxIndex = max(boneAddress, addedBoneMaxIndex)
	
	print("Highest bone index added to ArrayMesh %s" % addedBoneMaxIndex)
	
	var boneNumber = addedBoneMaxIndex + 1

	data[ArrayMesh.ARRAY_BONES] = newBones
	data[ArrayMesh.ARRAY_WEIGHTS] = newWeights
	
	var array_mesh = ArrayMesh.new()
	array_mesh.add_surface_from_arrays(Mesh.PRIMITIVE_TRIANGLES, data)

	var surface_tool = SurfaceTool.new()
	surface_tool.create_from(array_mesh, 0)
	surface_tool.generate_normals()
	
	print("Adding bones and skin count: %s" % boneNumber)

	for i in range(boneNumber):
		self.skeleton.add_bone("%s" % i)
		# Set initial resting position based on the assumed bone location and bone increment.
		var bonePosition = bonesIncrement * i - halfHeight
		var transform = Transform3D(Vector3(1, 0, 0), Vector3(0, 1, 0), Vector3(0, 0, 1), Vector3(0, bonePosition, 0))
		self.skeleton.set_bone_rest(i, transform)
	
		if i > 0:
			self.skeleton.set_bone_parent(i, i - 1)

	print("Setting skin and skeleton")

	self.meshInstance.skin = self.skeleton.create_skin_from_rest_transforms()
	self.meshInstance.mesh = surface_tool.commit()
```

**Comment by fire** [comment]
Here's a project for easier test. Running the project will generate a tscn.

[Skeleton Game Project.zip](https://github.com/godotengine/godot-proposals/files/10974682/Skeleton.Game.Project.zip)

**Comment by nvanfleet** [comment]
Thanks @fire  I will check that out. 

From what I can tell the mesh vertexes are created in the cylinder with a center position of zero. So the locations of the bones (I assume) should be placed in accordance with that, so I want bone 0 to be placed at -3 and the the top bone +3. I can be very wrong about that part though since the lack of docs. I've tried playing with those values and it's not 100% clear possibly because of some errors in Godot since I see that values changing sometimes causes weirdness and I have to restart. 

I did find some issues with my code though, I was referencing more bones than I was creating and a few other small things. I think I have to learn more about bones to be sure I did it right but I do have a pretty warped matrix at this point when I rotate the bones (!).
